server/service.go

```diff
@@ -243,44 +250,75 @@ func (m *BoostService) handleStatus(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
+type (
+	payloadByValidator map[relay.ValidatorPublicKey]types.SignedValidatorRegistration
+	payloadsByRelay    map[relay.Entry][]types.SignedValidatorRegistration
+)
+
+func (v payloadsByRelay) add(relays relay.List, payload types.SignedValidatorRegistration) {
+	for _, r := range relays {
+		v[r] = append(v[r], payload)
+	}
+}
+
 // handleRegisterValidator - returns 200 if at least one relay returns 200, else 502
 func (m *BoostService) handleRegisterValidator(w http.ResponseWriter, req *http.Request) {
 	log := m.log.WithField("method", "registerValidator")
-	log.Debug("registerValidator")
 
-	payload := []types.SignedValidatorRegistration{}
+	var payload []types.SignedValidatorRegistration
 	if err := DecodeJSON(req.Body, &payload); err != nil {
 		m.respondError(w, http.StatusBadRequest, err.Error())
 		return
 	}
 
 	ua := UserAgent(req.Header.Get("User-Agent"))
-	log = log.WithFields(logrus.Fields{
+	log.WithFields(logrus.Fields{
 		"numRegistrations": len(payload),
 		"ua":               ua,
-	})
+	}).Info("registering validators")
 
-	relayRespCh := make(chan error, len(m.relays))
+	validatorPayloads := make(payloadByValidator, len(payload))
+	for _, p := range payload {
+		pubKey := p.Message.Pubkey.String()
+		validatorPayloads[pubKey] = p
+	}
 
-	for _, relay := range m.relays {
-		go func(relay RelayEntry) {
-			url := relay.GetURI(pathRegisterValidator)
-			log := log.WithField("url", url)
+	relayPayloads := make(payloadsByRelay, len(m.relayConfigurator.AllRelays()))
+	for pubKey, p := range validatorPayloads {
+		relays := m.relayConfigurator.RelaysForProposer(pubKey)
+		if len(relays) < 1 {
+			log.Warnf("there are no relays specified for %s", pubKey)
+		}
+		relayPayloads.add(relays, p)
+	}
 
-			_, err := SendHTTPRequest(context.Background(), m.httpClientRegVal, http.MethodPost, url, ua, payload, nil)
-			relayRespCh <- err
+	wg := new(sync.WaitGroup)
+	relayErrCh := make(chan error, len(relayPayloads))
+
+	for r, payloads := range relayPayloads {
+		wg.Add(1)
+
+		go func(r relay.Entry, payloads []types.SignedValidatorRegistration) {
+			defer wg.Done()
+			relayURL := r.GetURI(pathRegisterValidator)
+			_, err := SendHTTPRequest(context.Background(), m.httpClientRegVal, http.MethodPost, relayURL, ua, payloads, nil)
+			relayErrCh <- err
 			if err != nil {
-				log.WithError(err).Warn("error calling registerValidator on relay")
+				log.WithField("url", relayURL).WithError(err).Warn("error calling registerValidator on relay")
 				return
 			}
-		}(relay)
+		}(r, payloads)
 	}
 
 	go m.sendValidatorRegistrationsToRelayMonitors(payload)
 
-	for i := 0; i < len(m.relays); i++ {
-		respErr := <-relayRespCh
-		if respErr == nil {
+	go func() {
+		wg.Wait()
+		close(relayErrCh)
+	}()
+
+	for err := range relayErrCh {
+		if err == nil {
 			m.respondOK(w, nilResponse)
```

Comment:
so this returns OK on the first successful response from any relay.

does this affect the other requests? should it wait for all to complete?

Reply:
The behaviour is exactly the same as it was before the changes.